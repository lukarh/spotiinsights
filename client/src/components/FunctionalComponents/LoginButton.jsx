import { useState } from "react";
import axios from 'axios';
import Button from '@mui/material/Button';
import React from 'react';

const LOGIN_BUTTON_LABEL = "Login to Spotify"

const LoginButton = () => {
    const [isProcessing, setIsProcessing] = useState(false)

    const handleLogin = async () => {
        setIsProcessing(true)

		try {
            // const baseURL = window.location.origin
            // const response = await axios.get(`/auth/login`, { withCredentials: true })
			const response = await axios.get('http://localhost:5000/auth/login', { withCredentials: true })

            if (response.status === 200) {
                const { redirectURL } = response.data
                window.location = redirectURL
                setIsProcessing(false)
            } else {
                setIsProcessing(false)
            }

		} catch (error) {
            setIsProcessing(false)
		}
	}

    return (
        <div className="padding-2-rem" data-testid="login-button">
            <Button className="auth-btn" variant="contained" onClick={handleLogin} disabled={isProcessing} >
                {LOGIN_BUTTON_LABEL}
            </Button>
        </div>
    )
}

export default LoginButton;