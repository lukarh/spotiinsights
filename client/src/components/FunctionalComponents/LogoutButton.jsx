import axios from "axios";
import Button from "@mui/material/Button";

const LOGOUT_BUTTON_LABEL = "Logout"

const LogoutButton = () => {
    const handleLogout = async () => {
        try {
            // const response = await axios.get(`/auth/logout`, { withCredentials: true })
			const response = await axios.get('http://localhost:5000/auth/logout', { withCredentials: true })

            if (response.status === 200) {
                const { redirectURL } = response.data
                window.location.href = redirectURL
            } 

		} catch (error) {
            // do nothing with error for now
		}
    }
    
    return (
        <div className="flex-center">
            <Button className="logout-btn" variant="contained" color="success" onClick={handleLogout}>
                {LOGOUT_BUTTON_LABEL}
            </Button>
        </div>
    )
}

export default LogoutButton;