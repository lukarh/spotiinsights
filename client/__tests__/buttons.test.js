import React from 'react';
import axios from 'axios';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { act, render, cleanup, fireEvent } from "@testing-library/react";

import LoginButton from '../src/components/FunctionalComponents/LoginButton';
import LogoutButton from '../src/components/FunctionalComponents/LogoutButton';
import RecommendButton from '../src/components/FunctionalComponents/RecommendButton';

jest.mock('axios') 

describe('Login Button Component', () => {
    const loginButtonLabel = "Login to Spotify"
    let hrefMock = jest.fn()

    beforeEach(() => {
        jest.clearAllMocks()

        delete window.location
        window.location = { href: '' }
    })

    afterEach(() => {
        cleanup()
        window.location = { href: hrefMock }
    })

    test('should render the login button with correct label', () => {
        const { getByText } = render(<LoginButton />)
        const loginButton = getByText(loginButtonLabel)
        expect(loginButton).toBeInTheDocument()
    })

    test('should call API login endpoint and get a response', async () => {
        const mockResponse = { status: 200, data: { redirectURL: 'https://www.google.com' }}
        axios.get.mockResolvedValue(mockResponse)

        const { getByText } = render(<LoginButton />)
        const loginButton = getByText(loginButtonLabel)

        await act(async () => {
            fireEvent.click(loginButton)
        })
        
        expect(axios.get).toHaveBeenCalledTimes(1)
        expect(axios.get).toHaveBeenCalledWith('http://localhost:5000/auth/login', { withCredentials: true })
        expect(mockResponse.data).toHaveProperty('redirectURL')
        expect(mockResponse.status).toBe(200)
    })

    test('should call API login endpoint and fail', async () => {
        const mockResponse = { status: 500, data: {} }
        const message = "Failed to start Spotify OAuth."
        axios.get.mockRejectedValueOnce(new Error(message))

        const { getByText } = render(<LoginButton />)
        const loginButton = getByText(loginButtonLabel)

        await act(async () => {
            fireEvent.click(loginButton)
        })

        expect(axios.get).toHaveBeenCalledTimes(1)
        expect(axios.get).toHaveBeenCalledWith('http://localhost:5000/auth/login', { withCredentials: true })
        expect(mockResponse.data).not.toHaveProperty('redirectURL')
        expect(mockResponse.status).toBe(500)
    })

    test('should match snapshot', () => {
        const tree = renderer.create(<LoginButton />).toJSON()
        expect(tree).toMatchSnapshot()
    })

})

describe('Logout Button Component', () => {
    const logoutButtonLabel = "Logout"

    beforeEach(() => {
        jest.clearAllMocks()
    })

    afterEach(() => {
        cleanup()
    })

    test('should render the logout button with correct label', () => {
        const { getByText } = render(<LogoutButton />)
        const logoutButton = getByText(logoutButtonLabel)
        expect(logoutButton).toBeInTheDocument()
    })

    test('should call API logout endpoint and get a response', async () => {
        const mockResponse = { status: 200, data: { redirectURL: 'https://www.google.com' }}
        axios.get.mockResolvedValue(mockResponse)

        const { getByText } = render(<LogoutButton />)
        const logoutButton = getByText(logoutButtonLabel)

        await act(async () => {
            fireEvent.click(logoutButton)
        })
        
        expect(axios.get).toHaveBeenCalledTimes(1)
        expect(axios.get).toHaveBeenCalledWith('http://localhost:5000/auth/logout', { withCredentials: true })
        expect(mockResponse.data).toHaveProperty('redirectURL')
        expect(mockResponse.status).toBe(200)
    })

    test('should call API logout endpoint and fail', async () => {
        const mockResponse = { status: 500, data: {} }
        const message = "Failed to logout of session."
        axios.get.mockRejectedValueOnce(new Error(message))

        const { getByText } = render(<LogoutButton />)
        const logoutButton = getByText(logoutButtonLabel)

        await act(async () => {
            fireEvent.click(logoutButton)
        })

        expect(axios.get).toHaveBeenCalledWith('http://localhost:5000/auth/logout', { withCredentials: true })
        expect(mockResponse.data).not.toHaveProperty('redirectURL')
        expect(mockResponse.status).toBe(500)
    })

    test('should match snapshot', () => {
        const tree = renderer.create(<LogoutButton />).toJSON()
        expect(tree).toMatchSnapshot()
    })

})

describe('Recommend Button Component', () => {
    const recommendButtonLabel = "Get Song Recommendations"

    beforeEach(() => {
        jest.clearAllMocks()
    })

    afterEach(() => {
        cleanup()
    })

    const getSongRecommendations = async (event) => {
        try {
            const searchResults = await axios.get(
                `http://localhost:5000/api/spotify/song-recommendations`, { withCredentials: true }
            )
        } catch (error) {
            // do nothing with the error 
        }
    }

    test('should render the recommend songs button with correct label', () => {
        
        const { getByRole } = render(<RecommendButton />)
        const recommendButton = getByRole('button')

        expect(recommendButton).toBeInTheDocument()
    })

    test('should call API recommendations endpoint and get a response', () => {
        const mockResponse = { status: 200, data: { items: [{ id: 1, track: "hello" }, { id: 2, track: "bye" }] } }
        axios.get.mockResolvedValue(mockResponse)

        const { getByRole } = render(<RecommendButton handleSubmit={getSongRecommendations}/>)
        const recommendButton = getByRole('button')

        act(() => {
            fireEvent.click(recommendButton)
        })

        expect(axios.get).toHaveBeenCalledTimes(1)
        expect(axios.get).toHaveBeenCalledWith('http://localhost:5000/api/spotify/song-recommendations', { withCredentials: true })
        expect(mockResponse.data).toHaveProperty('items')
        expect(mockResponse.status).toBe(200)
    })

    test('should call API recommendations endpoint and fail', () => {
        const mockResponse = { status: 500, data: {} }
        const message = "Failed to get recommended songs."
        axios.get.mockRejectedValueOnce(new Error(message))

        const { getByText } = render(<RecommendButton handleSubmit={getSongRecommendations}/>)
        const recommendButton = getByText(recommendButtonLabel)

        act(() => {
            fireEvent.click(recommendButton)
        })

        expect(axios.get).toHaveBeenCalledTimes(1)
        expect(axios.get).toHaveBeenCalledWith('http://localhost:5000/api/spotify/song-recommendations', { withCredentials: true })
        expect(mockResponse.data).not.toHaveProperty("items")
        expect(mockResponse.status).toBe(500)
    })

    test('should match snapshot', () => {
        const tree = renderer.create(<RecommendButton />).toJSON()
        expect(tree).toMatchSnapshot()
    })

})

