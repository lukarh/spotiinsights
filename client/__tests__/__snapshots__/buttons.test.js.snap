// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Button Component should match snapshot 1`] = `
<div
  className="padding-2-rem"
  data-testid="login-button"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium auth-btn css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Login to Spotify
  </button>
</div>
`;

exports[`Logout Button Component should match snapshot 1`] = `
<div
  className="flex-center"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium logout-btn css-ke5b6m-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Logout
  </button>
</div>
`;

exports[`Recommend Button Component should match snapshot 1`] = `
<div
  className="flex-column text-align-center padding-2-rem"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium recommend-btn css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Get Song Recommendations
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <small
    className="error-text padding-1-rem"
  />
</div>
`;
